tools/uninstall_opte.sh

```diff
@@ -0,0 +1,111 @@
+#
+# Small tool to _uninstall_ OPTE and the xde kernel driver and ONU bits.
+#
+# This should bring the system back to the stock Helios bits.
+
+set -e
+set -u
+set -x
+
+MARKER=/etc/opt/oxide/NO_INSTALL
+if [[ -f "$MARKER" ]]; then
+    echo "This system has the marker file $MARKER, aborting." >&2
+    exit 1
+fi
+
+if [[ "$(uname)" != "SunOS" ]]; then
+    echo "This script is intended for Helios only"
+    exit 1
+fi
+
+if [[ $(id -u) -ne 0 ]]; then
+    echo "This must be run as root"
+    exit 1
+fi
+
+# Remove these publishers if they exist
+PUBLISHERS=("on-nightly" "helios-netdev")
+
+# The stock Helios publisher and incorporation base information to revert to
+HELIOS_PUBLISHER="helios-dev"
+STOCK_CONSOLIDATION="pkg://$HELIOS_PUBLISHER/consolidation/osnet/osnet-incorporation"
+
+# Remove a publisher in $1, assuming it matches a few basic assumptions
+function remove_publisher {
+    local PUBLISHER="$1"
+    local LINE="$(pkg publisher | grep "$PUBLISHER" | tr -s ' ')"
+    if [[ -z "$LINE" ]]; then
+        echo "Publisher \"$PUBLISHER\" does not exist, ignoring"
+        return 0
+    fi
+    local ORIGIN="$(echo "$LINE" | cut -d ' ' -f 5)"
+    if [[ -z "$ORIGIN" ]]; then
+        echo "Cannot determine origin URI for publisher \"$PUBLISHER\""
+        exit 1
+    fi
+    if [[ "${ORIGIN:0:7}" != "file://" ]]; then
+        echo "Expected a file origin URI for publisher \"$PUBLISHER\""
+        exit 1
+    fi
+
+    # Unset the publisher, but do not do anything with the file.
+    echo "Unsetting publisher \"$PUBLISHER\""
+    pkg unset-publisher "$PUBLISHER"
+}
+
+# Install stock incorporation from helios-dev, pushing the publisher to the top
+function to_stock_helios {
+    local CONSOLIDATION="$1"
+    echo "Installing stock Helios kernel and networking bits"
+    echo "provided by the consolidation package:"
+    echo "\"$CONSOLIDATION\""
+    echo "Note that this may entail an upgrade, depending on"
+    echo "how your Helios system is configured."
+    pkg set-publisher --sticky --search-first "$HELIOS_PUBLISHER"
+    pkg install --no-refresh -v "$CONSOLIDATION"
+}
+
+# If helios-dev exists, echo the full osnet-incorporation package we'll be
+# installing to. If it does _not_ exist, fail.
+function ensure_helios_publisher_exists {
+    pkg publisher "$HELIOS_PUBLISHER" > /dev/null || \
+        echo "No \"$HELIOS_PUBLISHER\" publisher exists on this system!"
+    CONSOLIDATION="$(pkg list --no-refresh -H -af "$STOCK_CONSOLIDATION"@latest || echo "")"
+    if [[ -z "$CONSOLIDATION" ]]; then
+        echo "No osnet-incorporation package exists on this system,"
+        echo "so we cannot determine the exact package to install"
+        echo "to revert to stock Helios. You may need to update your"
+        echo "helios-dev publisher to refresh that publishers list of"
+        echo "available packages, with \"pkg refresh helios-dev\""
+        exit 1
+    fi
+    echo "$CONSOLIDATION" | tr -s ' ' | cut -d ' ' -f 1,3 | tr ' ' '@'
```

Comment:
Is this trying to concatenate the package with the version? On my machine it seems like it should be `-f 1,2` in that case

Reply:
So, if the publisher providing a package is not the first in the search order, it's listed in parentheses after the package name. From `man pkg`:

> If the publisher from which the package is installed (or available, if not
> installed) is not the first in the publisher search order, then the
> publisher name is listed in parentheses after the package name.

In this case, we've previously verified that `helios-dev` is not publisher that provides the `osnet-incorporation` package, so it should be listed like:

```
bnaecker@feldspar : ~ $ pkg list --no-refresh -H pkg://helios-dev/consolidation/osnet/osnet-incorporation@latest
consolidation/osnet/osnet-incorporation (helios-dev) 0.5.11-1.0.21067           im-
```

So we're munging our way to:

```
consolidation/osnet/osnet-incorporation@0.5.11-1.0.21067
```

At least, I _believe_ this is correct. I've tried it several times now, going back and forth between stock and xde-compatible Helios. But of course let me know if it does not work as intended!

Reply:
Ah, that explains it!